core/src/main/java/nl/nn/adapterframework/collection/CollectorPipeBase.java

```diff
@@ -0,0 +1,134 @@
+import nl.nn.adapterframework.pipes.FixedForwardPipe;
+import nl.nn.adapterframework.stream.Message;
+import nl.nn.adapterframework.util.ClassUtils;
+
+/**
+ * Base class for pipes that can collect items, such as multipart messages and zip archives.
+ *
+ * @author Niels Meijer
+ *
+ * @param <C> Collector instance
+ * @param <P> parts that are added to the collector
+ */
+public abstract class CollectorPipeBase<C extends ICollector<P>, P> extends FixedForwardPipe {
+
+	public enum Action {
+		/** To initiate a new collection */
+		OPEN,
+		/** Add an item to to an existing collection */
+		WRITE,
+		/** Combination of WRITE and CLOSE: Add an item to to an existing collection, then finalize the collection */
+		LAST,
+		/** Finalize the collection */
+		CLOSE;
+	}
+
+	/**
+	 * Collectors must be opened before you can write content to them, after it's closed you can no longer write to it.
+	 * @ff.default WRITE
+	 */
+	private @Getter @Setter Action action=Action.WRITE;
+
+	/** Session key used to refer to collection. Must be specified with another value if multiple CollectorPipes are active at the same time in the same session
+	  * @ff.default collection
+	 */
+	private @Getter @Setter String collectionName = "collection";
+
+	@Override
+	public void configure() throws ConfigurationException {
+		super.configure();
+
+		if(StringUtils.isBlank(collectionName)) {
+			throw new ConfigurationException("collectionName may not be blank");
+		}
+	}
+
+	@Override
+	public PipeRunResult doPipe(Message input, PipeLineSession session) throws PipeRunException {
+		try {
+			Message result = doAction(getAction(), input, session);
+			return new PipeRunResult(getSuccessForward(), result);
+		} catch (CollectionException e) {
+			throw new PipeRunException(this, "unable to ["+getAction()+"]", e);
+		}
+	}
+
+	protected @Nullable Collection<C, P> getCollection(PipeLineSession session) throws CollectionException {
+		@SuppressWarnings({ "unchecked", "rawtypes" })
+		Collection<C, P> collection = (Collection) session.get(getCollectionName());
+		if (collection == null && getAction() != Action.OPEN) {
+			throw new CollectionException("cannot find collection under key ["+getCollectionName()+"]");
+		}
+		return collection;
+	}
+
+	protected final @Nullable Message doAction(Action action, Message input, PipeLineSession session) throws CollectionException, PipeRunException {
+		try (CloseableThreadContext.Instance ctc = CloseableThreadContext.put("action", action.name())) {
+			Collection<C, P> collection = getCollection(session);
+			switch(action) {
+				case OPEN: {
+					if(collection != null) {
+						throw new CollectionException("collection [" + getCollectionName() + "] is already open, unable to create a new one");
+					}
+					collection = new Collection<>(createCollector());
+					session.scheduleCloseOnSessionExit(collection, ClassUtils.nameOf(this));
+					session.put(getCollectionName(), collection);
+					break; //should also be able to write?
```

Comment:
Should writing from OPEN then be optional?
Needs an extra parameter in that case to decide if to break, or fall thru to WRITE?

Let's decide and then remove this comment.

Reply:
I think it's better not to. I would be nice but allows parameters to be set on the OPEN action. I don't think this extra complexity is worth it. The LAST action already allows the combining of 2 actions, so you will always need a minimum of 2 actions to create a collection.